File: android-app/src/main/java/fr/cph/chicago/parser/BusStopCsvProcessor.java
package fr.cph.chicago.parser;

import com.annimon.stream.Stream;
import com.univocity.parsers.common.CommonSettings;
import com.univocity.parsers.common.Format;
import com.univocity.parsers.common.ParsingContext;
import com.univocity.parsers.common.processor.RowProcessor;

import java.util.ArrayList;
import java.util.List;

import fr.cph.chicago.entity.BusStop;
import fr.cph.chicago.entity.Position;
import io.realm.Realm;

class BusStopCsvProcessor implements RowProcessor {

    private List<BusStop> rows;
    private Realm realm;

    @Override
    public void processStarted(final ParsingContext context) {
        rows = new ArrayList<>(12000);
        realm = Realm.getDefaultInstance();
    }

    /**
     * Stores the row extracted by the parser into a list.
     *
     * @param row     the data extracted by the parser for an individual record. Note that:
     *                <ul>
     *                <li>it will never by null. </li>
     *                <li>it will never be empty unless explicitly configured using {@link CommonSettings#setSkipEmptyLines(boolean)}</li>
     *                <li>it won't contain lines identified by the parser as comments. To disable comment processing set {@link Format#setComment(char)} to '\0'</li>
     *                </ul>
     * @param context A contextual object with information and controls over the current state of the parsing process
     */
    @Override
    public void rowProcessed(final String[] row, final ParsingContext context) {
        final int stopId = Integer.parseInt(row[0]); // stop_id
        final int stopCode = row[1] == null ? -1 : Integer.parseInt(row[1]); // stop_code
        if (stopCode != -1) {
            final String stopName = row[2]; // stop_name
            final String stopDesc = row[3]; // stop_desc

            final double latitude = Double.parseDouble(row[4]);// stop_lat
            final double longitude = Double.parseDouble(row[5]);// stop_lon

            final BusStop busStop = new BusStop();
            busStop.setId(stopId);
            busStop.setName(stopName);
            busStop.setDescription(stopDesc);
            final Position position = new Position();
            position.setLatitude(latitude);
            position.setLongitude(longitude);
            busStop.setPosition(position);

            rows.add(busStop);
        }
    }

    @Override
    public void processEnded(final ParsingContext context) {
        realm.beginTransaction();
        Stream.of(rows).forEach(busStop -> realm.copyToRealm(busStop));
        realm.commitTransaction();
        realm.close();
    }
}

File: android-app/src/main/java/fr/cph/chicago/parser/BusStopCsvProcessor.kt
package fr.cph.chicago.parser

import com.univocity.parsers.common.CommonSettings
import com.univocity.parsers.common.Format
import com.univocity.parsers.common.ParsingContext
import com.univocity.parsers.common.processor.RowProcessor
import fr.cph.chicago.entity.BusStop
import fr.cph.chicago.entity.Position
import io.realm.Realm
import org.apache.commons.lang3.StringUtils
import java.util.*

internal class BusStopCsvProcessor : RowProcessor {

    private val rows: MutableList<BusStop> = ArrayList(12000)
    private val realm: Realm = Realm.getDefaultInstance()

    override fun processStarted(context: ParsingContext) {
    }

    /**
     * Stores the row extracted by the parser into a list.

     * @param row     the data extracted by the parser for an individual record. Note that:
     * *
     * *                 * it will never by null.
     * *                 * it will never be empty unless explicitly configured using [CommonSettings.setSkipEmptyLines]
     * *                 * it won't contain lines identified by the parser as comments. To disable comment processing set [Format.setComment] to '\0'
     * *
     * *
     * @param context A contextual object with information and controls over the current state of the parsing process
     */
    override fun rowProcessed(row: Array<String?>, context: ParsingContext) {
        if (row[1] != null) {
            val latitude = row[4]!!.toDouble() // stop_lat
            val longitude = row[5]!!.toDouble() // stop_lon
            val busStop = BusStop(
                id = row[0]!!.toInt(),
                name = row[2]!!,
                description = if (row[3] == null) StringUtils.EMPTY else row[3]!!,
                position = Position(latitude, longitude)
            )
            rows.add(busStop)
        }
    }

    override fun processEnded(context: ParsingContext) {
        realm.beginTransaction()
        rows.forEach { busStop -> realm.copyToRealm(busStop) }
        realm.commitTransaction()
        realm.close()
    }
}
